### DISCUSSION TITLE
I'm starting implementing a compiler plugin for the MoonBit language

### DISCUSSION THREAD

**Discussion by shamiao** [capnproto/capnproto#2063] [Ideas]
## About the language

[The MoonBit language](https://www.moonbitlang.com/) is a new programming language which compiles to WebAssembly and focuses on front-end and Edge computing. The language features and syntax are similar to Go and Rust, and it balances various factors (difficulty, WASM binary size, runtime performance, FFI) very well. 

MoonBit language is developed by a team from _Digital Infrastructure Innovation_ of _[International Digital Economy Academy (IDEA)](https://www.idea.edu.cn) (Shenzhen)_ leading by [Hongbo Zhang](https://github.com/bobzhang), who has built the ReScript language. It's currently in the early days of beta. 

## About me

I'm an independence developer having rich experience in the Rust language. I once involved the translation of _The Rust Programming Language_ (Chinese paperback) and now writing the official tutorial book of MoonBit. 

## Usage of capnp in MoonBit

Capnp will be mainly used for FFI, passing structured data across the bondage between the host language and WASM sandbox. Supporting capnp makes the following data flow available:

```
remote server <---network rpc---> my host environment by rust or c <---memcpy from/to wasm sandbox---> my script by moonbit
```

## Current development plan

1. the plugin name is both `capnpc-mbt` and `capnpc-moonbit`. 
2. the good old MoonBit is statically typed and compiles AOT. Traditional code generation is the only need. 
3. I'd like to bootstrap with json in the first version. my plugin will do this to the stdin from `capnp` compiler: 

```
capnp convert binary:json schema.capnp CodeGeneratorRequest
(executed by plugin, stdin is the capnp passed to capnpc-mbt)
```

**Comment by jonahbeckford** [comment]
I'm sure others are much better suited to answer about your high-level design and plan, but you may find my older post at https://www.mail-archive.com/capnproto@googlegroups.com/msg01891.html relevant for FFI + capnp.

(I remember Hongbo posted about MoonBit last year on the OCaml discussion boards. Good to see the progress!)
